ee/tabby-webserver/src/cron/github.rs

```diff
@@ -0,0 +1,119 @@
+use std::{collections::HashSet, sync::Arc};
+
+use anyhow::Result;
+use chrono::Utc;
+use juniper::ID;
+use octocrab::{models::Repository, GitHubError, Octocrab};
+use tracing::warn;
+
+use crate::schema::github_repository_provider::{
+    GithubRepositoryProvider, GithubRepositoryProviderService,
+};
+
+pub async fn refresh_all_repositories(
+    service: Arc<dyn GithubRepositoryProviderService>,
+) -> Result<()> {
+    for provider in service
+        .list_github_repository_providers(vec![], None, None, None, None)
+        .await?
+    {
+        let start = Utc::now();
+        refresh_repositories_for_provider(service.clone(), provider.id.clone()).await?;
+        service
+            .delete_outdated_github_provided_repositories(provider.id, start)
+            .await?;
+    }
+    Ok(())
+}
+
+async fn refresh_repositories_for_provider(
+    service: Arc<dyn GithubRepositoryProviderService>,
+    provider_id: ID,
+) -> Result<()> {
+    let cached_repositories: HashSet<_> = service
```

Comment:
by using `upsert_github_provided_repository`, you can avoid creating this HashSet